### PR TITLE
v6 support

### PR DESCRIPTION
**Pull request by cmin764**

### CODE REVIEW

File: argus/scenarios/cloud.py [hunk 1]
```diff
@@ -144,33 +159,42 @@ def get_network_interfaces(self):
...
+            net_details = net_client.show_network(network_id)["network"]
+            nic = dict.fromkeys(util.NETWORK_KEYS)
+            for subnet_id in net_details["subnets"]:
+                details = net_client.show_subnet(subnet_id)["subnet"]
+
+                # The network interface should follow the format found under
+                # `windows.InstanceIntrospection.get_network_interfaces`
+                # method or `argus.util.NETWORK_KEYS` model.
+                v6switch = details["ip_version"] == 6
+                v6suffix = lambda: "6" if v6switch else ""
```
Comment (PCManticore) [inline argus/scenarios/cloud.py]: Isn't `v6suffis = "6" if v6switch else ""` enough? Why does it have to be a callable?

File: argus/scenarios/cloud.py [hunk 2]
```diff
@@ -144,33 +159,42 @@ def get_network_interfaces(self):
...
+                nic["dhcp"] = details["enable_dhcp"]
+                nic["dns" + v6suffix()] = details["dns_nameservers"]
+                nic["gateway" + v6suffix()] = details["gateway_ip"]
```
Comment (PCManticore) [inline argus/scenarios/cloud.py]: gateway_ip is the V6 gateway?

File: argus/scenarios/cloud.py [hunk 3]
```diff
@@ -144,33 +159,42 @@ def get_network_interfaces(self):
...
+                nic["netmask" + v6suffix()] = util.cidr2netmask(
```
Comment (PCManticore) [inline argus/scenarios/cloud.py]: I prefer a separate function, cidr2netmask_ipv6. Then you'll choose the function depending if you are working with IPV6 or IPV4.

File: etc/argus.conf [hunk 4]
```diff
@@ -49,6 +49,7 @@ DEFAULT.config_drive_cdrom=false
 
 config_file = /etc/nova/nova.conf
 DEFAULT.flat_injected=True
+DeFAULT.use_ipv6=True
```
Comment (PCManticore) [inline etc/argus.conf]: De?

File: argus/introspection/cloud/windows.py [hunk 5]
```diff
@@ -68,6 +68,19 @@ def _escape_path(path):
     return path
 
 
+def _getip_v4_v6(ip_string):
```
Comment (PCManticore) [inline argus/introspection/cloud/windows.py]: ip_string is a space separated string of ips?

File: argus/introspection/cloud/windows.py [hunk 6]
```diff
@@ -286,12 +299,21 @@ def get_network_interfaces(self):
...
+            gateways, gateways6 = _getip_v4_v6(details[2])
+            netmasks, netmasks6 = _getip_v4_v6(details[3])
+            dnses, dnses6 = _getip_v4_v6(details[4])
             nic = {
                 "mac": details[0],
-                "address": details[1].split(" ")[0],
-                "gateway": details[2].split(" ")[0],
-                "netmask": details[3].split(" ")[0],
-                "dns": details[4].split(" "),
+                "address": addresses[0] if addresses else None,
```
Comment (PCManticore) [inline argus/introspection/cloud/windows.py]: Why don't you set them conditionally?

if V6: set address6 & friends
else: set address4 & friends.

File: argus/scenarios/cloud.py (hunk 2 above)
Comment (cmin764) [inline argus/scenarios/cloud.py]: Yes, they are on the same name, but they contain different type values depending on `ip_version`.

File: argus/introspection/cloud/windows.py (hunk 5 above)
Comment (cmin764) [inline argus/introspection/cloud/windows.py]: Yes.

File: argus/introspection/cloud/windows.py (hunk 6 above)
Comment (cmin764) [inline argus/introspection/cloud/windows.py]: There is no switch between address versions, they are all sorted in 2 groups then chosen by custom corresponding indexes.

Comment (PCManticore) [inline argus/introspection/cloud/windows.py]: Yes there is. If it's IPV6, then you are retrieving the last index, else the first one. Couldn't you do:

if gateway6: # I know I have IPV6
   set ipv6 metadata
else:
   set ipv4

File: argus/introspection/cloud/windows.py [hunk 7]
```diff
@@ -68,6 +68,19 @@ def _escape_path(path):
     return path
 
 
+def _getips_v4_v6(ips_as_string):
```
Comment (PCManticore) [inline argus/introspection/cloud/windows.py]: As I can see, you are using only the first and the last entry, so why does it have to be a list? Return a namedtuple, which can be used as this:

ip.gateway6
ip.gateway
ip.address
ip.address6

Cleaner, no need to know the position of the specific IP.

File: argus/scenarios/cloud.py [hunk 8]
```diff
@@ -144,33 +159,45 @@ def get_network_interfaces(self):
...
+            net_details = net_client.show_network(network_id)["network"]
+            nic = dict.fromkeys(util.NETWORK_KEYS)
+            for subnet_id in net_details["subnets"]:
+                details = net_client.show_subnet(subnet_id)["subnet"]
+
+                # The network interface should follow the format found under
+                # `windows.InstanceIntrospection.get_network_interfaces`
+                # method or `argus.util.NETWORK_KEYS` model.
+
+                # pylint: disable=cell-var-from-loop
```
Comment (PCManticore) [inline argus/scenarios/cloud.py]: False positive or actual bug on your side?

File: argus/introspection/cloud/windows.py [hunk 9]
```diff
@@ -68,6 +68,27 @@ def _escape_path(path):
...
+    for ip in ips:
+        if not ip:
+            continue
+        if "." in ip and ":" not in ip:
+            ips_v4.append(ip)
+        else:
+            ips_v6.append(ip)
+    if not single:
+        return ips_v4, ips_v6
+    return (ips_v4[0] if ips_v4 else None,
```
Comment (PCManticore) [inline argus/introspection/cloud/windows.py:88]: I dislike this solution. You control the output through two separate flags and the result is something unmanageable. Here's something better:
1. use something such as the following code in network_details.ps1. Basically, you're annotating each line with whatever field you're wanting.
   
      echo $sep
      [string]::Format("MAC {0}", $nic.MACAddress -join " ")
      [string]::Format("ip {0}", ($nic.IPAddress -join " "))
      [string]::Format("DefaultIpGateway {0}", ($nic.DefaultIPGateway -join " "))
      [string]::Format("Subnet {0}", ($nic.IPSubnet -join " "))
2. parse this result and use a collections.namedtuple for each component, the result being something like this:

result = network_details.mac.ipv4
            network_details.mac.ipv6
            network_details.ip.ipv4
            network_details.ip.ipv6
            etc.

Which means that you'll be interested in parsing each line. Split it by space, if it has two entries, the first ip v4, the second is ip v6. How does this sound?

**Comment by cmin764** [comment]
Reopened here: https://github.com/PCManticore/argus-ci/pull/69

**Pull request closed without merging**